carnot/carnot_vote_aggregation.py

```diff
@@ -0,0 +1,421 @@
+        # in this case the node will prepare a QC from votes it has received.
+        # assert len(votes) == self.overlay.super_majority_threshold(self.id)
+        assert all(self.overlay.is_member_of_subtree(self.id, vote.voter) for vote in votes)
+        assert all(vote.block == block.id() for vote in votes)
+        assert self.highest_voted_view < block.view
+
+        # Create a QC based on committee membership
+        qc = self.build_qc(block.view, block, None)  # if self.overlay.is_member_of_root_committee(self.id) else None
+
+        # Create a new vote
+        vote = Vote(
+            block=block.id(),
+            voter=self.id,
+            view=block.view,
+            qc=qc
+        )
+
+        # Update the highest voted view
+        self.highest_voted_view = max(self.highest_voted_view, block.view)
+
+        # After block verification, votes are sent to committee members.
+        # When a QC is formed from 2/3rd of subtree votes, it's forwarded to the parent committee.
+        # If a Type 1 timeout occurs, a QC is built from available votes and QCs and sent to the parent.
+        # Subsequent votes are forwarded to the parent committee members.
+        if self.overlay.is_member_of_root_committee():
+            recipient = self.overlay.leader(block.view + 1)
+        else:
+            recipient = self.overlay.my_committee(self.id)
+
+        # Return a Send event to the appropriate recipient
+        return Send(to=recipient, payload=vote)
+
+    # NewView msgs are not needed anymore
+    def build_qc(self, view: View, block: Optional[Block], timeouts: Optional[Set[Timeout]]) -> Qc:
+        # unhappy path
+        if timeouts:
+            timeouts = list(timeouts)
+            return AggregateQc(
+                qcs=[msg.high_qc.view for msg in timeouts],
+                highest_qc=max(timeouts, key=lambda x: x.high_qc.view).high_qc,
+                view=timeouts[0].view
+            )
+        # happy path
+        return StandardQc(
+            view=view,
+            block=block.id()
+        )
+
+    # A node initially forwards a vote or qc from its subtree to its parent committee. There can be two instances this
+    # can happen: 1: If a node forms a QC qc from votes and QCs it receives from its subtree such that the total number of votes in the qc is at two-third of votes from the subtree, then
+    # it forwards this QC to the parent committee members or a subset of parent committee members.
+    # 2: After sending the qc any additional votes are forwarded to the parent committee members or a subset of parent committee members.
+    # 3: After type 1 timeout a node builds a QC from arbitrary number of votes+QCs it has received, building a QC qc such that total number of votes in qc is less
+    # than the two-thirds of the number of the nodes in the sub-tree.
+
+    def forward_vote_qc(self, vote: Optional[Vote] = None, qc: Optional[Qc] = None) -> Optional[Event]:
```

Comment:
I do not really understand why to treat different voting if reached threshold or not. When, if I understood correctly. What we could end up is just a vote why a* threshold of votes/qcs or without it* ?

Reply:
Something like:

```mermaid
flowchart TD
    A[Listen votes] -->B[Got timeout]
    A --> C[Got threshold]
    C -->|Enough| D[Build QC]
    B -->|Not enough| D
```

Reply:
It avoids rebuilding overlay even if threshold is not reached. If number of malicious nodes in a committee prevents from QC formation then after t1 timeout a committee member will forward the QC with smaller number of votes. This mechanism with forwarding any additional votes will prevent failure if some committees have more malicious nodes than 1/3rd.